File: solutions/code-carl/ch08_greedy/T17_LC0968_MinCameraCover.java
/**
 * Practice template for:
 * - LeetCode 968. Binary Tree Cameras
 *   https://leetcode.cn/problems/binary-tree-cameras/
 */
/**
 * @author Cheng Yihao
 * @version 1.0
 * @date 2025/4/7 16:51
 * @comment
 */
public class T17_LC0968_MinCameraCover {
    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode() {
        }

        TreeNode(int val) {
            this.val = val;
        }

        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }

    int count = 0;
    public int minCameraCover(TreeNode root) {
        int rootVal = traversal(root);
        if (rootVal == 0) {
            count++;
        }
        return count;
    }

    public int traversal(TreeNode root) {
        // 0 未被覆盖
        // 1 有摄像头
        // 2 被覆盖
        if (root == null) {
            return 2;
        }
        int left = traversal(root.left);
        int right = traversal(root.right);
        if (left == 0 || right == 0) {
            count++;
            return 1;
        } else if (left == 1 || right == 1) {
            return 2;
        } else if (left == 2 && right == 2) {
            return 0;
        } else {
            return -1;
        }
    }

}

File: templates/python/code-carl/ch08_greedy/t17_lc0968_min_camera_cover.py
"""LeetCode 968. Binary Tree Cameras
https://leetcode.cn/problems/binary-tree-cameras/
"""

from __future__ import annotations

from typing import Optional

from tree_node import TreeNode


class Solution:

    def minCameraCover(self, root: Optional[TreeNode]) -> int:
        pass


if __name__ == "__main__":
    def build_tree(nodes):
        if not nodes:
            return None
        values = list(nodes)
        root_val = values.pop(0)
        if root_val is None:
            return None
        root = TreeNode(root_val)
        queue = [root]
        while values and queue:
            node = queue.pop(0)
            if values:
                left_val = values.pop(0)
                if left_val is not None:
                    node.left = TreeNode(left_val)
                    queue.append(node.left)
            if values:
                right_val = values.pop(0)
                if right_val is not None:
                    node.right = TreeNode(right_val)
                    queue.append(node.right)
        return root


    solution = Solution()
    assert solution.minCameraCover(build_tree([0, 0, None, 0, 0])) == 1
    assert solution.minCameraCover(build_tree([0, 0, None, 0, None, 0, None, None, 0])) == 2
    print("All tests passed.")
